x-pack/plugins/enterprise_search/server/lib/ml_inference_pipeline/get_inference_pipelines.ts

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IngestPipeline, IngestProcessorContainer } from '@elastic/elasticsearch/lib/api/types';
+import { ElasticsearchClient } from '@kbn/core/server';
+import { MlTrainedModels } from '@kbn/ml-plugin/server';
+
+import { getMlModelConfigsForModelIds } from '../indices/fetch_ml_inference_pipeline_processors';
+
+/**
+ * Gets all ML inference pipelines. Redacts trained model IDs in those pipelines which reference
+ * a model inaccessible in the current Kibana space.
+ * @param esClient the Elasticsearch Client to use to fetch the errors.
+ * @param trainedModelsProvider ML trained models provider.
+ */
+export const getMlInferencePipelines = async (
+  esClient: ElasticsearchClient,
+  trainedModelsProvider: MlTrainedModels | undefined
+): Promise<Record<string, IngestPipeline>> => {
+  if (!trainedModelsProvider) {
+    return Promise.reject(new Error('Machine Learning is not enabled'));
+  }
+
+  // Fetch all ML inference pipelines
+  const fetchedInferencePipelines = await esClient.ingest.getPipeline({
+    id: 'ml-inference-*',
+  });
+
+  // Fetch all trained models from the current Kibana space, and identify those that have their
+  // model_id redacted
+  const trainedModels = await getMlModelConfigsForModelIds(esClient, trainedModelsProvider);
+  const redactedModelIds = Object.values(trainedModels)
+    .filter((modelConfig) => !modelConfig.modelId)
+    .map((modelConfig) => modelConfig.trainedModelName);
+
+  // Process pipelines: check if the model_id is one of the redacted ones, if so, redact it in the
+  // result as well
+  const inferencePipelinesResult: Record<string, IngestPipeline> = {};
+  Object.keys(fetchedInferencePipelines).forEach((name) => {
+    const inferencePipeline = fetchedInferencePipelines[name];
+
+    inferencePipelinesResult[name] = isModelIdRedacted(inferencePipeline, redactedModelIds)
+      ? redactModelId(inferencePipeline)
+      : inferencePipeline;
+  });
+
+  return Promise.resolve(inferencePipelinesResult);
+};
+
+/**
+ * Convenience function that finds the inference processor in a given ingest pipeline.
+ * @param pipeline the pipeline.
+ */
+function getInferenceProcessor(pipeline: IngestPipeline): IngestProcessorContainer | undefined {
```

Comment:
Although it might not happen unless a user edits the pipeline, what would happen if an ingest pipeline has more than one `inference` processor... as-is, it will only return the first one... is this the desired behaviour?

Reply:
Ah, good point. Yes, in that case we need to apply the redaction to all `inference` processors. I'll modify the code to iterate on `processors`.